File: tools/tui/src/lib/device-link.ts
import { randomBytes } from "node:crypto";
import { createServer, IncomingMessage, ServerResponse } from "node:http";
import { spawn } from "node:child_process";
import { URL } from "node:url";

interface BrowserLoginOptions {
  webBaseUrl: string;
  timeoutMs?: number;
  onLog?: (message: string) => void;
}

interface BrowserLoginResult {
  token: string;
}

interface CallbackBody {
  token?: unknown;
  nonce?: unknown;
}

function normalizeBaseUrl(baseUrl: string): string {
  return baseUrl.endsWith("/") ? baseUrl.slice(0, -1) : baseUrl;
}

function readJsonBody(request: IncomingMessage): Promise<CallbackBody> {
  return new Promise((resolve, reject) => {
    let raw = "";

    request.on("data", (chunk) => {
      raw += chunk.toString();
      if (raw.length > 32_000) {
        reject(new Error("Request body too large"));
      }
    });

    request.on("end", () => {
      try {
        resolve(JSON.parse(raw) as CallbackBody);
      } catch {
        reject(new Error("Invalid JSON payload"));
      }
    });

    request.on("error", reject);
  });
}

function sendJson(response: ServerResponse, status: number, body: unknown): void {
  response.statusCode = status;
  response.setHeader("Content-Type", "application/json");
  response.setHeader("Access-Control-Allow-Origin", "*");
  response.setHeader("Access-Control-Allow-Headers", "Content-Type");
  response.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
  response.end(JSON.stringify(body));
}

function tryOpenBrowser(url: string): boolean {
  try {
    if (process.platform === "darwin") {
      const child = spawn("open", [url], {
        detached: true,
        stdio: "ignore",
      });
      child.unref();
      return true;
    }

    if (process.platform === "win32") {
      const child = spawn("cmd", ["/c", "start", "", url], {
        detached: true,
        stdio: "ignore",
      });
      child.unref();
      return true;
    }

    const child = spawn("xdg-open", [url], {
      detached: true,
      stdio: "ignore",
    });
    child.unref();
    return true;
  } catch {
    return false;
  }
}

export async function runBrowserLoginFlow(
  options: BrowserLoginOptions
): Promise<BrowserLoginResult> {
  const webBaseUrl = normalizeBaseUrl(options.webBaseUrl);
  const timeoutMs = options.timeoutMs ?? 180_000;
  const nonce = randomBytes(16).toString("hex");

  return await new Promise<BrowserLoginResult>((resolve, reject) => {
    const server = createServer(async (request, response) => {
      if (request.method === "OPTIONS") {
        response.statusCode = 204;
        response.setHeader("Access-Control-Allow-Origin", "*");
        response.setHeader("Access-Control-Allow-Headers", "Content-Type");
        response.setHeader("Access-Control-Allow-Methods", "POST, OPTIONS");
        response.end();
        return;
      }

      if (request.method !== "POST" || request.url !== "/callback") {
        sendJson(response, 404, { error: "Not found" });
        return;
      }

      try {
        const body = await readJsonBody(request);
        if (body.nonce !== nonce) {
          sendJson(response, 400, { error: "Invalid nonce" });
          return;
        }
        if (typeof body.token !== "string" || body.token.length === 0) {
          sendJson(response, 400, { error: "Token is required" });
          return;
        }

        sendJson(response, 200, { ok: true });
        settle(null, { token: body.token });
      } catch (error) {
        const message =
          error instanceof Error ? error.message : "Invalid callback payload";
        sendJson(response, 400, { error: message });
      }
    });

    let settled = false;

    const settle = (error: Error | null, result?: BrowserLoginResult) => {
      if (settled) return;
      settled = true;
      clearTimeout(timeoutId);
      server.close();
      if (error) {
        reject(error);
      } else if (result) {
        resolve(result);
      }
    };

    const timeoutId = setTimeout(() => {
      settle(new Error("Authentication timed out"));
    }, timeoutMs);

    server.on("error", (error) => {
      settle(error instanceof Error ? error : new Error("Callback server failed"));
    });

    server.listen(0, "127.0.0.1", () => {
      const address = server.address();
      if (!address || typeof address === "string") {
        settle(new Error("Failed to start callback server"));
        return;
      }

      const callback = new URL("http://127.0.0.1");
      callback.port = String(address.port);
      callback.pathname = "/callback";

      const browserUrl = `${webBaseUrl}/tui/link?callback=${encodeURIComponent(
        callback.toString()
      )}&nonce=${encodeURIComponent(nonce)}`;

      options.onLog?.("Waiting for browser authentication...");
      options.onLog?.(`Open this URL if browser does not open: ${browserUrl}`);
      if (!tryOpenBrowser(browserUrl)) {
        options.onLog?.("Could not auto-open browser. Please open the URL manually.");
      }
    });
  });
}

File: tools/tui/internal/tui/device_link.go
package tui

import (
	"context"
	"crypto/rand"
	"encoding/hex"
	"encoding/json"
	"errors"
	"fmt"
	"net"
	"net/http"
	"net/url"
	"os/exec"
	"runtime"
	"strings"
	"time"
)

type BrowserLoginOptions struct {
	WebBaseURL string
	Timeout    time.Duration
}

type BrowserLoginResult struct {
	Token string
}

type callbackBody struct {
	Token string `json:"token"`
	Nonce string `json:"nonce"`
}

type callbackResult struct {
	Token string
	Err   error
}

func randomNonce() (string, error) {
	b := make([]byte, 16)
	if _, err := rand.Read(b); err != nil {
		return "", err
	}
	return hex.EncodeToString(b), nil
}

func tryOpenBrowser(link string) {
	var cmd *exec.Cmd
	switch runtime.GOOS {
	case "darwin":
		cmd = exec.Command("open", link)
	case "windows":
		cmd = exec.Command("cmd", "/c", "start", "", link)
	default:
		cmd = exec.Command("xdg-open", link)
	}
	_ = cmd.Start()
}

func sendJSON(w http.ResponseWriter, status int, body any) {
	w.Header().Set("Content-Type", "application/json")
	w.Header().Set("Access-Control-Allow-Origin", "*")
	w.Header().Set("Access-Control-Allow-Headers", "Content-Type")
	w.Header().Set("Access-Control-Allow-Methods", "POST, OPTIONS")
	w.WriteHeader(status)
	_ = json.NewEncoder(w).Encode(body)
}

func RunBrowserLoginFlow(options BrowserLoginOptions) (BrowserLoginResult, error) {
	if options.Timeout <= 0 {
		options.Timeout = 3 * time.Minute
	}

	nonce, err := randomNonce()
	if err != nil {
		return BrowserLoginResult{}, err
	}

	ln, err := net.Listen("tcp", "127.0.0.1:0")
	if err != nil {
		return BrowserLoginResult{}, err
	}
	defer ln.Close()

	resultCh := make(chan callbackResult, 1)

	mux := http.NewServeMux()
	mux.HandleFunc("/callback", func(w http.ResponseWriter, r *http.Request) {
		if r.Method == http.MethodOptions {
			sendJSON(w, http.StatusNoContent, map[string]any{})
			return
		}
		if r.Method != http.MethodPost {
			sendJSON(w, http.StatusNotFound, map[string]string{"error": "Not found"})
			return
		}

		var body callbackBody
		if err := json.NewDecoder(http.MaxBytesReader(w, r.Body, 32_000)).Decode(&body); err != nil {
			sendJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid JSON payload"})
			return
		}

		if body.Nonce != nonce {
			sendJSON(w, http.StatusBadRequest, map[string]string{"error": "Invalid nonce"})
			return
		}
		if strings.TrimSpace(body.Token) == "" {
			sendJSON(w, http.StatusBadRequest, map[string]string{"error": "Token is required"})
			return
		}

		sendJSON(w, http.StatusOK, map[string]bool{"ok": true})

		select {
		case resultCh <- callbackResult{Token: body.Token}:
		default:
		}
	})

	server := &http.Server{Handler: mux}
	go func() {
		if err := server.Serve(ln); err != nil && !errors.Is(err, http.ErrServerClosed) {
			select {
			case resultCh <- callbackResult{Err: err}:
			default:
			}
		}
	}()

	callbackURL := url.URL{
		Scheme: "http",
		Host:   ln.Addr().String(),
		Path:   "/callback",
	}

	base := NormalizeBaseURL(options.WebBaseURL)
	if base == "" {
		base = "http://localhost:3000"
	}

	browserURL := fmt.Sprintf(
		"%s/tui/link?callback=%s&nonce=%s",
		base,
		url.QueryEscape(callbackURL.String()),
		url.QueryEscape(nonce),
	)
	tryOpenBrowser(browserURL)

	timer := time.NewTimer(options.Timeout)
	defer timer.Stop()

	select {
	case result := <-resultCh:
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()
		_ = server.Shutdown(ctx)
		if result.Err != nil {
			return BrowserLoginResult{}, result.Err
		}
		return BrowserLoginResult{Token: result.Token}, nil
	case <-timer.C:
		ctx, cancel := context.WithTimeout(context.Background(), 2*time.Second)
		defer cancel()
		_ = server.Shutdown(ctx)
		return BrowserLoginResult{}, errors.New("authentication timed out")
	}
}
